cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -426,11 +438,56 @@ public Set<ProgramId> filterProgramsExistence(Collection<ProgramId> programIds)
   }
 
 
-  public void writeApplication(String namespaceId, String appId, String versionId, ApplicationSpecification spec)
-    throws IOException {
-    writeApplicationSerialized(namespaceId, appId, versionId, GSON.toJson(new ApplicationMeta(appId, spec)));
+  public void writeApplication(String namespaceId, String appId, String versionId, ApplicationSpecification spec,
+                               ChangeDetail change) throws IOException {
+    writeApplicationSerialized(namespaceId, appId, versionId, GSON.toJson(new ApplicationMeta(appId, spec, change)),
```

Comment:
@albertshau So I pass in `ChangeDetail ` here to `ApplicationMeta` which is serialized. I saw your previous comment about not passing in these fields - I am not sure how to do that - since ChangeDetail is non-nullable

Reply:
doesn't it have to be nullable because of the pre-existing rows?